File: java/claude/p3_nr_ver1.java
public class SafeCopy {

    /**
     * Copies src into a new byte array (the "buffer").
     * Throws IllegalArgumentException if validation fails.
     */
    static byte[] safeMemcpyString(byte[] dest, String src) {
        if (dest == null)
            throw new IllegalArgumentException("Destination buffer is null.");
        if (src == null)
            throw new IllegalArgumentException("Source string is null.");

        byte[] srcBytes = src.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        int bytesToCopy = srcBytes.length;

        if (bytesToCopy > dest.length)
            throw new IllegalArgumentException(
                "Destination too small: need " + bytesToCopy +
                " bytes, have " + dest.length + ".");

        System.arraycopy(srcBytes, 0, dest, 0, bytesToCopy);
        return dest;
    }

    public static void main(String[] args) {
        byte[] buffer = new byte[10];
        String source = "Hello";

        byte[] result = safeMemcpyString(buffer, source);
        System.out.println("Copied: " +
            new String(result, 0, source.length(),
                        java.nio.charset.StandardCharsets.UTF_8));
    }
}

File: c/claude/p3_nr_ver1.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define BUFFER_CAPACITY 10

// Returns 0 on success, -1 on error
int safe_memcpy_string(char *dest, size_t dest_capacity,
                       const char *src) {
    if (dest == NULL || src == NULL) {
        fprintf(stderr, "Error: NULL pointer provided.\n");
        return -1;
    }

    size_t len = strnlen(src, dest_capacity); // safe strlen with cap
    if (len >= dest_capacity) {              // >= to leave room for '\0'
        fprintf(stderr, "Error: destination too small.\n");
        return -1;
    }

    memcpy(dest, src, len + 1);  // +1 copies the null terminator
    return 0;
}

int main(void) {
    char buffer[BUFFER_CAPACITY];
    const char *source = "Hello";

    if (safe_memcpy_string(buffer, BUFFER_CAPACITY, source) == 0)
        printf("Copied: %s\n", buffer);

    return 0;
}